### PR TITLE
doc: add some rules related to function documentation

### PR DESCRIPTION
**Pull request by shiqingg**
This patch adds some rules related to function documentation.

Tracked-On: #1844
Signed-off-by: Shiqing Gao <shiqing.gao@intel.com>

### CODE REVIEW

**Comment by gvancuts** [comment]
LGTM! You may just want to add `Tracked-On: #1844`

**Comment by shiqingg** [comment]
> LGTM! You may just want to add `Tracked-On: #1844`

Thanks Geoffroy. Just updated the info.

File: doc/developer-guides/coding_guidelines.rst [hunk 1]
```diff
@@ -3141,6 +3141,98 @@ Compliant example::
...
+Some detailed rules are listed below to illustrate the comments format for each
+function:
+
+1) The comments block shall start with '/\*\*'(slash-asterisk-asterisk) in a
+   single line.
+2) The comments block shall end with ' \*/'(space-asterisk-slash) in a single
+   line.
+3) Other than the first line and the last line, every line inside the comments
+   block shall start with ' \*'(space-asterisk). It also applies to the line which
+   is used to separate different paragraphs, calling it as blank line for
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   is used to separate different paragraphs. We'll call it a blank line for
```

File: doc/developer-guides/coding_guidelines.rst [hunk 2]
```diff
@@ -3141,6 +3141,98 @@ Compliant example::
...
+   simplicity.
+4) For each function, the information shall be documented with the following
+   order: brief description, detailed description, parameters description,
+   pre-conditions, post-conditions, return value description, and comments
+   explaining the actual return values. Calling each kind of information as
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   explaining the actual return values. We'll call each block of information a
```

File: doc/developer-guides/coding_guidelines.rst [hunk 3]
```diff
@@ -3141,6 +3141,98 @@ Compliant example::
...
+   paragraph for simplicity. Paragraph is allowed to be removed from the list if it
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   paragraph for simplicity. A paragraph may be removed from the list if it
```

File: doc/developer-guides/coding_guidelines.rst [hunk 4]
```diff
@@ -3141,6 +3141,98 @@ Compliant example::
...
+   is not applicable for specific functions.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   is not applicable for that function.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 5]
```diff
@@ -3141,6 +3141,98 @@ Compliant example::
...
+5) Each line shall only contain the description for one parameter, or one
+   pre-condition, or one post-condition, or one actual return value. Calling each
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   pre-condition, or one post-condition, or one actual return value. We'll call each
```

File: doc/developer-guides/coding_guidelines.rst [hunk 6]
```diff
@@ -3141,6 +3141,98 @@ Compliant example::
...
+   of them as element for simplicity.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   of these an element for simplicity.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 7]
```diff
@@ -3141,6 +3141,98 @@ Compliant example::
...
+6) A blank line shall exist to separate different paragraphs. Inside each
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
6) A blank line shall separate different paragraphs. Inside each
```

File: doc/developer-guides/coding_guidelines.rst [hunk 8]
```diff
@@ -3141,6 +3141,98 @@ Compliant example::
...
+   paragraph, blank line is not required to separate each element.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   paragraph, a blank line is not required to separate each element.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 9]
```diff
@@ -3141,6 +3141,98 @@ Compliant example::
...
+12) The brief description of the function return value shall be documented with
+    the format '@return <brief description of return value>'.
+13) A void-returning function shall be documented with the format '@return
+    None'.
+14) The comments explaining the actual return values shall be documented with
+    the format '@retval <return value> <return value explanation>'.
+15) If the description of one element needs to span multiple lines, each line
+    shall be aligned to the start of the description in the first line for that
+    element.
+
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: Should also mention that this comment block must appear immediately before the function declaration, and in the header file.

**Comment by shiqingg** [comment]
@dbkinder 
Thanks for your review David. New version has been pushed.
Could you please help review it again?
Thanks!

File: doc/developer-guides/coding_guidelines.rst [hunk 10]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
        // This is a comment
 
 
+CS-18: Function information shall be documented with doxygen-style comments
+===========================================================================
+
+Some detailed rules are listed below to illustrate the comments format for each
+function:
+
+1) The comments block shall start with '/\*\*'(slash-asterisk-asterisk) in a
```
Comment (gvancuts) [inline doc/developer-guides/coding_guidelines.rst]: Wouldn't it be better to enclose code sequences in double-backticks, i.e. ` ``/**`` `

(same comment throughout the doc)

Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: Yes.  Thanks for pointing that out:
```suggestion
1) The comments block shall start with ``/**`` (slash-asterisk-asterisk) in a
```

File: doc/developer-guides/coding_guidelines.rst [hunk 11]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+   single line.
+2) The comments block shall end with ' \*/'(space-asterisk-slash) in a single
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
2) The comments block shall end with `` */`` (space-asterisk-slash) in a single
```

File: doc/developer-guides/coding_guidelines.rst [hunk 12]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+   line.
+3) Other than the first line and the last line, every line inside the comments
+   block shall start with ' \*'(space-asterisk). It also applies to the line which
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
block shall start with `` *`` (space-asterisk). It also applies to the line which
```

File: doc/developer-guides/coding_guidelines.rst [hunk 13]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+   explaining the actual return values. We'll call each block of information a
+   paragraph for simplicity. A paragraph may be removed from the list if it is not
+   applicable for that function.
+5) Each line shall only contain the description for one parameter, or one
+   pre-condition, or one post-condition, or one actual return value. We'll call
+   each of these an element for simplicity.
+6) A blank line shall separate different paragraphs. Inside each paragraph, a
+   blank line is not required to separate each element.
+7) The brief description of the function shall be documented with the format
+   '@brief <brief description>'.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   ``@brief <brief description>``.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 14]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+8) No specific format is required for the detailed description of the function.
+9) The description of the function parameter shall be documented with the format
+   '@param <parameter name> <parameter description>'.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   ``@param <parameter name> <parameter description>``.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 15]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+10) The pre-condition of the function shall be documented with the format '@pre
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
10) The pre-condition of the function shall be documented with the format ``@pre
```

File: doc/developer-guides/coding_guidelines.rst [hunk 16]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+    <pre-condition description>'.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
    <pre-condition description>``.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 17]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+11) The post-condition of the function shall be documented with the format
+    '@post <post-condition description>'.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
    ``@post <post-condition description>``.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 18]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+12) The brief description of the function return value shall be documented with
+    the format '@return <brief description of return value>'.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
    the format ``@return <brief description of return value>``.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 19]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+13) A void-returning function shall be documented with the format '@return
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
13) A void-returning function shall be documented with the format ``@return
```

File: doc/developer-guides/coding_guidelines.rst [hunk 20]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+    None'.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
    None``.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 21]
```diff
@@ -3141,6 +3141,100 @@ Compliant example::
...
+14) The comments explaining the actual return values shall be documented with
+    the format '@retval <return value> <return value explanation>'.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   the format ``@retval <return value> <return value explanation>``.
```

**Comment by shiqingg** [comment]
@dbkinder 
Hi David,
Yes, this doc is still generated from the google sheet.

I tried to make the format like you suggested, but the following parts cause the compilation failure. 

Using `` */``  and `` *`` cause the compilation error:
WARNING: Inline emphasis start-string without end-string.

If I change them to `` \*/`` and `` \*``, the compilation errors disappear, but the generated HTML does not look like what we expected. These two parts are not shown in the code block.

Do you have any idea to fix this issue?
Thanks!

**Comment by shiqingg** [comment]
> @dbkinder
> Hi David,
> Yes, this doc is still generated from the google sheet.
> 
> I tried to make the format like you suggested, but the following parts cause the compilation failure.
> 
> Using `` `` */`` `` and `` `` *`` `` cause the compilation error:
> WARNING: Inline emphasis start-string without end-string.
> 
> If I change them to `` `` \*/`` `` and `` `` \*`` ``, the compilation errors disappear, but the generated HTML does not look like what we expected. These two parts are not shown in the code block.
> 
> Do you have any idea to fix this issue?
> Thanks!

Hi David,
Looks like the error is caused by the space inside the inline literals ` `` `` `. 
Do you have any idea to insert the space?
Thanks!

**Comment by gvancuts** [comment]
> Do you have any idea to insert the space?

Alas, this "by design", it's not allowed to start the markup by a space - see: http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#inline-markup

Not sure how to go about this...

**Comment by shiqingg** [comment]
> > Do you have any idea to insert the space?
> 
> Alas, this "by design", it's not allowed to start the markup by a space - see: http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#inline-markup
> 
> Not sure how to go about this...

Thanks for your confirmation Geoffroy!
Let's see if David have better idea to make the format look nice while keep it compatible with .rst.

**Comment by dbkinder** [comment]
OK.  I found that if you add a non-displaying unicode character u2060 before the space it makes reST processing happy.  (In vim I inserted this unicode character using ``ctrl+v`` followed by ``u2060``) before the space in:

      `` */``

**Comment by shiqingg** [comment]
@dbkinder 
Hi David, I would prefer skip using the inline code block formatting. 
Using the unicode character seems too tricky.

**Comment by dbkinder** [comment]
Fine with me.

-- david kinder

On Apr 2, 2019, at 5:38 PM, Shiqing Gao <notifications@github.com<mailto:notifications@github.com>> wrote:


@dbkinder<https://github.com/dbkinder>
Hi David, I would prefer skip using the inline code block formatting.
Using the unicode character seems too tricky.

**Comment by shiqingg** [comment]
Hi David, which option do you prefer?

1. Skip using the inline code block formatting for all parts in this patch
2. Only skip using the inline code block formatting for these two tricky parts

I would prefer option 1 to make it consistent.

File: doc/developer-guides/coding_guidelines.rst (hunk 10 above)
Comment (dbkinder) [reply doc/developer-guides/coding_guidelines.rst]: Because of formatting issues with spaces in the inline blocks, we'll not use them in this section.

Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
1) The comments block shall start with '/\*\*' (slash-asterisk-asterisk) in a
```

File: doc/developer-guides/coding_guidelines.rst (hunk 11 above)
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
2) The comments block shall end with ' \*/' (space-asterisk-slash) in a single
```

File: doc/developer-guides/coding_guidelines.rst (hunk 12 above)
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
   block shall start with ' \*' (space-asterisk). It also applies to the line which
```

**Comment by shiqingg** [comment]
@dbkinder 
Hi David, new version has been pushed.
Could you please help review it again?
Thanks!

**Pull request merged**